File: src/design-pattern/factory-method.py
from abc import ABC, abstractmethod
from typing import List  # Abstract Base Class


# Abstract product
class AbstractBlock(ABC):
    @abstractmethod
    def produce(self):
        pass


# Concrete product A
class CircleBlock(AbstractBlock):
    def produce(self):
        print("Circle Block")


# Concrete product B
class SquareBlock(AbstractBlock):
    def produce(self):
        print("Square Block")


# Abstract factory
class AbstractBlockFactory(ABC):
    @abstractmethod
    def createBlock(self) -> AbstractBlock:
        pass


# Concrete Factory A
class CircleBlockFactory(AbstractBlockFactory):
    def createBlock(self) -> AbstractBlock:
        return CircleBlock()


# Concrete Factory B
class SquareBlockFactory(AbstractBlockFactory):
    def createBlock(self) -> AbstractBlock:
        return SquareBlock()


class BlockFactorySystem:
    def __init__(self) -> None:
        self.blocks: List[AbstractBlock] = []

    def produceBlocks(self, factory: AbstractBlockFactory, cnt: int):
        for _ in range(cnt):
            block = factory.createBlock()
            self.blocks.append(block)
            block.produce()

    def getBlocks(self):
        return self.blocks


if __name__ == "__main__":
    factorySystem = BlockFactorySystem()
    productionCnt = int(input().strip())

    circleBlockFactory = CircleBlockFactory()
    squareBlockFactory = SquareBlockFactory()

    for _ in range(productionCnt):
        blockType, cntStr = input().strip().split(" ")
        cnt = int(cntStr)

        if blockType == "Circle":
            factorySystem.produceBlocks(circleBlockFactory, cnt)
            continue

        if blockType == "Square":
            factorySystem.produceBlocks(squareBlockFactory, cnt)

File: src/design-pattern/factory-method.ts
// 简单工厂: 一个工厂方法创建所有具体产品
// 工厂方法: 一个工厂方法创建一个具体产品
// 抽象工厂: 一个工厂方法创建一类具体产品

// 工厂方法模式
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

abstract class AbstractBlock {
  abstract produce(): void;
}

class CircleBlock extends AbstractBlock {
  produce() {
    console.log("Circle Block");
  }
}

class SquareBlock extends AbstractBlock {
  produce() {
    console.log("Square Block");
  }
}

abstract class AbstractBlockFactory {
  abstract createBlock(): AbstractBlock;
}

class CircleBlockFactory extends AbstractBlockFactory {
  createBlock(): AbstractBlock {
    return new CircleBlock();
  }
}

class SquareBlockFactory extends AbstractBlockFactory {
  createBlock(): AbstractBlock {
    return new SquareBlock();
  }
}

class BlockFactorySystem {
  blocks: AbstractBlock[] = [];

  produceBlocks(factory: AbstractBlockFactory, cnt: number) {
    for (let i = 0; i < cnt; i++) {
      const block = factory.createBlock();
      this.blocks.push(block);
      block.produce();
    }
  }
}

let lineno = 0;
let productionCnt = 0;
const circleBlockFactory = new CircleBlockFactory();
const squareBlockFactory = new SquareBlockFactory();
const factorySystem = new BlockFactorySystem();

rl.on("line", (line) => {
  lineno++;
  if (lineno === 1) {
    productionCnt = Number.parseInt(line.trim());
    return;
  }

  const [blockType, cntStr] = line.trim().split(" ");
  const cnt = Number.parseInt(cntStr);

  if (blockType === "Circle") {
    factorySystem.produceBlocks(circleBlockFactory, cnt);
  }

  if (blockType === "Square") {
    factorySystem.produceBlocks(squareBlockFactory, cnt);
  }

  if (lineno === productionCnt + 1) {
    rl.close();
  }
});

// rl.on("pause", () => {
//   rl.close()
// })
